context("Utility functions")
test_that("test several utility functions", {
  x1<-plot_mrk_info(tetra.solcap, 1)
  x2<-plot_mrk_info(tetra.solcap, "solcap_snp_c2_41437")
  x3<-plot_mrk_info(tetra.solcap.geno.dist, 1)
  x4<-plot_mrk_info(tetra.solcap.geno.dist, "solcap_snp_c2_41437")
  expect_is(x1, "matrix")
  expect_is(x2, "matrix")
  expect_is(x3, "matrix")
  expect_is(x4, "matrix")
  expect_equal(get_LOD(solcap.err.map[[1]]),0)
  tpt<-est_pairwise_rf(make_seq_mappoly(tetra.solcap, 1:30))
  x1<-make_seq_mappoly(tetra.solcap, 1:20)
  x2<-make_seq_mappoly(tetra.solcap, 21:40)
  x1<-as.numeric(mappoly:::pos_twopt_est(x1))
  x2<-as.numeric(mappoly:::pos_twopt_est(x2))
  expect_equal(as.numeric(crossprod(x1,x2)), 1)
  M<-rf_list_to_matrix(tpt)
  expect_equal(round(sum(get_rf_from_mat(M$rec.mat), na.rm = TRUE), 6), 3.913633)
  expect_equal(get_w_m(6), 15)
  expect_error(get_w_m(0))
  expect_error(get_w_m(3))
  rm<-rev_map(maps.hexafake[[1]])
  expect_identical(sum(rm$maps[[1]]$seq.rf), sum(maps.hexafake[[1]]$maps[[1]]$seq.rf))
  a1<-sample_data(tetra.solcap.geno.dist, n = 20, type = "marker")
  a2<-sample_data(a1, n = 20, type = "individual")
  plot(a2)
  a3<-dist_prob_to_class(a2$geno)
  expect_equal(prod(dim(a3$geno.dose)), nrow(a3$geno))
  w1<-export_data_to_polymapR(hexafake)
  expect_equal(dim(w1) , c(1500, 302))
  expect_equal(sum(w1), 461530)
  w2<-export_data_to_polymapR(tetra.solcap)
  expect_equal(dim(w2) , c(3679, 162))
  expect_equal(sum(w2, na.rm = TRUE), 1058080)
  expect_equal(round(mf_h(20),6), 0.16484)
  expect_equal(round(mf_k(20),6), 0.189974)
  expect_equal(round(mf_m(20),6), 0.2)
  expect_equal(round(imf_h(.15),2), 17.83)
  expect_equal(round(imf_k(.15),2), 15.48)
  expect_equal(round(imf_m(.15),2), 15)
  expect_null(plot_compare_haplotypes(m = 6, 
                          hom.allele.p1 = maps.hexafake[[1]]$maps[[1]]$seq.ph$P[1:10],
                          hom.allele.q1 = maps.hexafake[[1]]$maps[[1]]$seq.ph$Q[1:10],
                          hom.allele.p2 = maps.hexafake[[1]]$maps[[1]]$seq.ph$P[1:10],
                          hom.allele.q2 = maps.hexafake[[1]]$maps[[1]]$seq.ph$Q[1:10]))
  expect_null(print_mrk(input.data = hexafake, mrks = "M_1"))
  expect_equal(sum(perm_pars(1:5)), 900)
  expect_equal(sum(perm_tot(1:5)), 1800)
  a<-sample_data(tetra.solcap.geno.dist, n=30, type = "marker")
  a<-sample_data(a, n=30)
  w<-update_missing(a, prob.thres = .7)
  expect_is(w, "mappoly.data")
  w2<-get_genomic_order(make_seq_mappoly(hexafake, "all"))
  w3<-round(as.numeric(crossprod(w2$seq.pos)/10e17),3)
  expect_equal(w3, 0.495)
})